### PR TITLE
Implement an "ordered" option in concurrent streams

### PR DESCRIPTION
**Pull request by harendra-kumar**
This is the equivalent of the Ahead style streams.

### CODE REVIEW

File: src/Streamly/Internal/Data/Stream/Async/Channel/Operations.hs [hunk 1]
```diff
@@ -281,6 +278,9 @@ _fromChannelD svar = D.Stream step FromSVarInit
     step _ (FromSVarLoop sv (ev : es)) = do
         case ev of
             ChildYield a -> return $ D.Yield a (FromSVarLoop sv es)
+            ChildStopChannel -> do
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/Async/Channel/Operations.hs:281]: Does this mean that this function was initially incorrect? It did not account for the exception `ChildStop`?

File: src/Streamly/Internal/Data/Stream/Async/Channel.hs [hunk 2]
```diff
@@ -163,7 +164,10 @@ appendGeneric newChan modifier stream1 stream2 = K.concatEffect action
 {-# INLINE appendWithK #-}
 appendWithK :: MonadAsync m =>
     (Config -> Config) -> K.Stream m a -> K.Stream m a -> K.Stream m a
-appendWithK = appendGeneric Append.newChannel
+appendWithK modifier stream1 stream2 =
+    if getOrdered (modifier defaultConfig)
+    then concatMapWithK modifier id (stream1 `K.cons` K.fromPure stream2)
+    else appendGeneric Append.newChannel modifier stream1 stream2
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/Async/Channel.hs:170]: When the `ordered` is not set in the modifier are both the branches here same?

Comment (adithyaov) [reply src/Streamly/Internal/Data/Stream/Async/Channel.hs:170]: If so, is `appendGeneric` more performant? Hench the `if` statement exists?

File: src/Streamly/Internal/Data/Stream/Async/Channel/Append.hs [hunk 3]
```diff
@@ -207,6 +213,727 @@ workLoopLIFOLimited qref sv winfo = run
...
+-- efficiently.
+--
+-- XXX Maybe we can start the ahead threads at a lower cpu and IO priority so
+-- that they do not hog the resources and hinder the progress of the threads in
+-- front of them.
+
+-- Left associated ahead expressions are expensive. We start a new SVar for
+-- each left associative expression. The queue is used only for right
+-- associated expression, we queue the right expression and execute the left.
+-- Thererefore the queue never has more than on item in it.
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/Async/Channel/Append.hs:467]: All this is also written above. This documentation is duplicated.

File: src/Streamly/Internal/Data/Stream/Async/Channel/Append.hs [hunk 4]
```diff
@@ -274,17 +1011,32 @@ getLifoSVar mrun cfg = do
...
+                    else enqueueLIFO sv q inner
+            , eagerDispatch = when eager $ void $ dispatchWorker 0 sv
+            , isWorkDone =
+                if inOrder
+                then isWorkDoneAhead sv aheadQ outH
+                else workDone sv
+            , isQueueDone =
+                if inOrder
+                then isQueueDoneAhead sv aheadQ
+                else workDone sv
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/Async/Channel/Append.hs:1032]: I've only reviewed this part properly. I assume most of the code is copied from the other module. If not, let know.

**Review by adithyaov — COMMENTED** [review]
Some comments.

File: src/Streamly/Internal/Data/Stream/Async/Channel/Operations.hs (hunk 1 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Stream/Async/Channel/Operations.hs:281]: Yes, should have handled the ChannelStop exception in the previous change.

File: src/Streamly/Internal/Data/Stream/Async/Channel.hs (hunk 2 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Stream/Async/Channel.hs:170]: Checked performance, it is the same. We can use concat itself for both cases. But in future if we want to use StreamD for fusion, the special append impl may be useful.

File: src/Streamly/Internal/Data/Stream/Async/Channel/Append.hs (hunk 3 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Stream/Async/Channel/Append.hs:467]: Yes, will remove.

**Pull request merged**